src/main/scopes/js/metrics/function-metric.ts

```diff
@@ -0,0 +1,120 @@
+ */
+import { JsScopeAttributes, NpmScopeAttributes } from '@ibm/telemetry-attributes-js'
+import type { ConfigSchema } from '@ibm/telemetry-config-schema'
+import type { Attributes } from '@opentelemetry/api'
+
+import { hash } from '../../../core/anonymize/hash.js'
+import { substitute } from '../../../core/anonymize/substitute.js'
+import type { Logger } from '../../../core/log/logger.js'
+import { PackageDetailsProvider } from '../../../core/package-details-provider.js'
+import { ScopeMetric } from '../../../core/scope-metric.js'
+import type { PackageData } from '../../npm/interfaces.js'
+import type { JsFunction, JsImport } from '../interfaces.js'
+
+/**
+ * JSX scope metric that generates a jsx.function individual metric for a given function.
+ */
+export class FunctionMetric extends ScopeMetric {
+  public override name = 'js.function' as const
+  private readonly jsFunction: JsFunction
+  private readonly matchingImport: JsImport
+  private readonly allowedArgumentStringValues: string[]
+  private readonly instrumentedPackage: PackageData
+
+  /**
+   * Constructs a TokenMetric.
+   *
+   * @param jsToken - Object containing token data to generate metric from.
+   * @param matchingImport - Import that matched the provided JsFunction in the file.
+   * @param instrumentedPackage - Data (name and version) pertaining to instrumented package.
+   * @param config - Determines which argument values to collect for.
+   * @param logger - Logger instance.
+   */
+  public constructor(
+    jsToken: JsFunction,
+    matchingImport: JsImport,
+    instrumentedPackage: PackageData,
+    config: ConfigSchema,
+    logger: Logger
+  ) {
+    super(logger)
+    this.jsFunction = jsToken
+    this.matchingImport = matchingImport
+    this.instrumentedPackage = instrumentedPackage
+
+    this.allowedArgumentStringValues =
+      config.collect.js?.functions?.allowedArgumentStringValues ?? []
+  }
+
+  /**
+   * Get all OpenTelemetry Attributes for this metric data point.
+   *
+   * @returns OpenTelemetry compliant attributes, anonymized and substituted where necessary.
+   */
+  public override get attributes(): Attributes {
+    const packageDetailsProvider = new PackageDetailsProvider(this.logger)
+
+    const {
+      owner: instrumentedOwner,
+      name: instrumentedName,
+      major: instrumentedMajor,
+      minor: instrumentedMinor,
+      patch: instrumentedPatch,
+      preRelease: instrumentedPreRelease
+    } = packageDetailsProvider.getPackageDetails(
+      this.instrumentedPackage.name,
+      this.instrumentedPackage.version
+    )
+
+    // convert arguments into a fake object to satisfy the substitute api
+    const argsObj: Record<string, unknown> = this.jsFunction.arguments.reduce(
+      (cur, val, index) => ({ ...cur, [index.toString()]: val }),
+      {}
+    )
+
+    const anonymizedArguments = substitute(
+      argsObj,
+      Object.keys(argsObj), // all keys are allowed
+      this.allowedArgumentStringValues
+    )
```

Comment:
Revisit this "hack", we'd want to improve the API for substitute so that it's able to take in just values (array) instead of objects, while maintaining all other current use cases